Tpetra: parallel kernels for CRS matrix addition (#1367)

brian-kelley opened the pull request:
-Makes TpetraExt::MatrixMatrix::add very thread scalable in most cases
-Kernels using Kokkos to locally add two Tpetra CrsMatrices
-One kernel for the sorted graph, matching col map case, and
 a more general one for unsorted graph and/or mismatched col map case
-Unit tests for both kernels

mhoemmen commented on packages/tpetra/core/ext/TpetraExt_MatrixMatrix_def.hpp:
```diff
@@ -479,6 +478,103 @@ void Add(
+      }
+      if(finalPass)
+      {
+        if(i == entryUnion.dimension_0() - 1)
```
Does anything ever use the last entry in `gtol`?  If not, we could skip this branch altogether.

brian-kelley replied:
I do use it once to get the size of the GView colmap.

mhoemmen replied:
How do you get that result back on host?  Do you do a `Kokkos::deep_copy` of a 0-D View, or just UVM it?  For the former, see Tpetra_Details_getEntryOnHost.hpp.

brian-kelley replied:
Good catch, it's going to be UVM but I forgot the fence().

mhoemmen commented on packages/tpetra/core/ext/TpetraExt_MatrixMatrix_def.hpp:
```diff
@@ -479,6 +478,103 @@ void Add(
+  //build local set of GIDs for constructing column map - the last entry in gtol is the total number of local cols
+  GView colmap("column map", gtol(gtol.dimension_0() - 1));
+  size_t localIter = 0;
+  for(size_t i = 0; i < entryUnion.dimension_0(); i++)
```
This implicitly assumes UVM.  If we ever want to get rid of the `CUDA_LAUNCH_BLOCKING` requirement, we need to put a fence (`execution_space::fence()`) before and after accessing `entryUnion` and `colmap` on host.

mhoemmen replied:
Also, could we reformulate this sequential host kernel as a `parallel_scan`?  We could have counted the total number of nonzero entries in `entryUnion`, when building `entryUnion` above.  I need to think about how this scan would look.

mhoemmen commented on packages/tpetra/core/ext/TpetraExt_MatrixMatrix_decl.hpp:
```diff
@@ -128,6 +128,130 @@ void Add(
+  typedef typename device_type::execution_space execution_space;
+  typedef typename crs_matrix_type::impl_scalar_type ISC;
+  typedef typename crs_matrix_type::local_matrix_type KCRS;
+  typedef typename KCRS::values_type::non_const_type ValuesArray;
```
Also, this needs to go in a "Details" inner namespace.

mhoemmen commented on packages/tpetra/core/ext/TpetraExt_MatrixMatrix_decl.hpp:
```diff
@@ -128,6 +128,130 @@ void Add(
+    const RowPtrsArray& Browptrs,
+    const ColIndsArray& Bcolinds, 
+    const ISC scalarB,
+    Teuchos::RCP<ValuesArray>& Cvals,
```
ValuesArray is just a Kokkos::View, right?  Why do you need to pass it by nonconst `RCP&`?  If you mean for the View to be an output argument, just pass in `ValuesArray&`.  Ditto for `Crowptrs` and `Ccolinds` below.

mhoemmen commented on packages/tpetra/core/ext/TpetraExt_MatrixMatrix_decl.hpp:
```diff
@@ -128,6 +128,130 @@ void Add(
+    const local_map_type& BcolMap,
+    GlobalOrdinal minGlobalCol,
+    GlobalOrdinal globalNumCols,
+    Teuchos::RCP<ValuesArray>& Cvals,
```
ValuesArray is just a Kokkos::View, right?  Why do you need to pass it by nonconst `RCP&`?  If you mean for the View to be an output argument, just pass in `ValuesArray&`.  Ditto for `Crowptrs` and `Ccolinds` below.

mhoemmen commented on packages/tpetra/core/ext/TpetraExt_MatrixMatrix_decl.hpp:
```diff
@@ -128,6 +128,130 @@ void Add(
+    const ColIndsArray& Bcolinds, 
+    const ISC scalarB,
+    GlobalOrdinal numGlobalCols,
+    Teuchos::RCP<ValuesArray>& Cvals,
```
ValuesArray is just a Kokkos::View, right?  Why do you need to pass it by nonconst `RCP&`?  If you mean for the View to be an output argument, just pass in `ValuesArray&`.  Ditto for `Crowptrs` and `Ccolinds` below.

mhoemmen commented on packages/tpetra/core/ext/TpetraExt_MatrixMatrix_decl.hpp:
```diff
@@ -128,6 +128,130 @@ void Add(
+    const GlobalColIndsArray& Bcolinds, 
+    const ISC scalarB,
+    GlobalOrdinal numGlobalCols,
+    Teuchos::RCP<ValuesArray>& Cvals,
```
ValuesArray is just a Kokkos::View, right?  Why do you need to pass it by nonconst `RCP&`?  If you mean for the View to be an output argument, just pass in `ValuesArray&`.  Ditto for `Crowptrs` and `Ccolinds` below.

mhoemmen commented on packages/tpetra/core/ext/TpetraExt_MatrixMatrix_decl.hpp:
```diff
@@ -128,6 +128,130 @@ void Add(
+    Teuchos::RCP<RowPtrsArray>& Crowptrs,
+    Teuchos::RCP<GlobalColIndsArray>& Ccolinds);
+
+  static Teuchos::RCP<map_type> getOptimizedColMap(GlobalOrdinal ncols, const GlobalColIndsArray& gids, ColIndsArray& lids, const Teuchos::RCP<const Teuchos::Comm<int>>& comm);
```
What does "optimized" mean here?  I'm not keen on functions with names like "fast" or "better."  You could instead say something like "makeMapWithLids" (it's a "maker" and not a "getter").

brian-kelley replied:
I meant "optimized" in the sense of owning the minimum set of global cols to cover the given entries. I could just call it makeColMapAndConvertGids since that's really what it does (and it only gets called in one place so it's OK if the name is long).

mhoemmen commented on packages/tpetra/core/ext/TpetraExt_MatrixMatrix_def.hpp:
```diff
@@ -479,6 +478,103 @@ void Add(
+    {
+      entryUnion(gids(i)) = 1;
+    }
+    ByteView& entryUnion;
```
Kokkos::View has shallow copy semantics.  If your functor holds onto Kokkos::View instances by nonconst reference, this functor won't work on GPUs.  A reference is a pointer, and it points to host memory, where the class instance lives.  The class instance _contains_ a GPU device allocation.  Just store the Kokkos::View by value in the functor.

mhoemmen commented on packages/tpetra/core/ext/TpetraExt_MatrixMatrix_def.hpp:
```diff
@@ -479,6 +478,103 @@ void Add(
+    {
+      lids(i) = gtol(gids(i));
+    }
+    const LView& gtol;
```
Kokkos::View has shallow copy semantics.  If your functor holds onto Kokkos::View instances by nonconst reference, this functor won't work on GPUs.  A reference is a pointer, and it points to host memory, where the class instance lives.  The class instance _contains_ a GPU device allocation.  Just store the Kokkos::View by value in the functor.  Also, "const View<T>" doesn't mean the same thing as "View<const T>".  You'll have to store these as "View" rather than "const View" anyway.

mhoemmen commented on packages/tpetra/core/ext/TpetraExt_MatrixMatrix_def.hpp:
```diff
@@ -521,60 +621,145 @@ add (const Scalar& alpha,
+  typedef typename AddKern::ColIndsArray ColIndsArray;
+  bool AGraphSorted = Aprime->getCrsGraph()->isSorted();
+  bool BGraphSorted = Bprime->getCrsGraph()->isSorted();
+  RCP<ValuesArray> vals;
```
See above.  Kokkos::View is a lot like Teuchos::ArrayRCP.  You wouldn't generally have an `RCP<ArrayRCP>`, would you?

mhoemmen commented on packages/tpetra/core/ext/TpetraExt_MatrixMatrix_def.hpp:
```diff
@@ -521,60 +621,145 @@ add (const Scalar& alpha,
+#ifdef HAVE_TPETRA_MMM_TIMINGS
+      MM = rcp(new TimeMonitor(*TimeMonitor::getNewTimer(prefix_mmm + std::string("Tpetra::Crs expertStaticFillComplete"))));
+#endif
+  C->expertStaticFillComplete(CDomainMap, CRangeMap, rcp(new import_type(CDomainMap, CcolMap)), rcp(new export_type(CrowMap, CRangeMap)), params);
```
This would have been a good opportunity to use the fused Import constructor, no?

mhoemmen commented on packages/tpetra/core/ext/TpetraExt_MatrixMatrix_def.hpp:
```diff
@@ -521,60 +621,145 @@ add (const Scalar& alpha,
+    CcolMap = Bprime->getColMap();
+  }
+#ifdef HAVE_TPETRA_MMM_TIMINGS
+      MM = rcp(new TimeMonitor(*TimeMonitor::getNewTimer(prefix_mmm + std::string("Tpetar::Crs constructor"))));
```
Please fix spelling error.  Also, I'm worried that the timing code creates a whole lot of std::string instances....

brian-kelley replied:
@mhoemmen Thanks for catching that. I will profile everything again to see how bad std::string is within the add() (I think it's going to be measurable but not worth avoiding).

mhoemmen commented on packages/tpetra/core/ext/TpetraExt_MatrixMatrix_def.hpp:
```diff
@@ -895,6 +1080,687 @@ void Add(
+//Given array containing entries per row, produce rowptrs
+//Used by all add kernels
+template<typename RowPtrType>
+struct PrefixSumFunctor
```
Couldn't we just use the existing function `Tpetra::Details::computeOffsetsFromCounts`?

brian-kelley replied:
I just did this (haven't checked in yet).

mhoemmen commented on packages/tpetra/core/ext/TpetraExt_MatrixMatrix_def.hpp:
```diff
@@ -895,6 +1080,687 @@ void Add(
+      }
+    }
+  }
+  RowPtrType& Crowcounts;
```
See above note on Kokkos::View needing to be stored by value in functors, not by reference.

mhoemmen commented on packages/tpetra/core/ext/TpetraExt_MatrixMatrix_def.hpp:
```diff
@@ -895,6 +1080,687 @@ void Add(
+    numEntries += Browlen - bi;
+    Crowcounts(i) = numEntries;
+  }
+  const RowPtrs& Arowptrs;
```
See above note on Kokkos::View needing to be stored by value in functors, not by reference.

mhoemmen commented on packages/tpetra/core/ext/TpetraExt_MatrixMatrix_def.hpp:
```diff
@@ -895,6 +1080,687 @@ void Add(
+{
+
+template<typename Ordinal, typename RowPtrs, typename ColInds>
+struct EntryCountingFunctor
```
Would it make sense to fuse this functor with the row offsets functor?  A single parallel scan should be able to take care of this.  It's true that this would require iterating over each row twice, but it could save kernel launch overhead and thus should help with strong scaling.

brian-kelley replied:
@mhoemmen I did this in the 3rd commit.

brian-kelley replied:
@mhoemmen Combined the entry count and prefix sum is about twice as slow as doing them separately. I think that's because the expensive part (scanning entries in rows and matching indices) is being executed on every non-final pass of scan. Before I was caching all the entry counts first and then prefix summing. I think I'll have to leave them separate.

mhoemmen commented on packages/tpetra/core/ext/TpetraExt_MatrixMatrix_def.hpp:
```diff
@@ -895,6 +1080,687 @@ void Add(
+      numInserted++;
+    }
+  }
+  const Values& Avals;
```
See above note on Kokkos::View needing to be stored by value in functors, not by reference.

mhoemmen commented on packages/tpetra/core/ext/TpetraExt_MatrixMatrix_def.hpp:
```diff
@@ -895,6 +1080,687 @@ void Add(
+  {
+    CrowcountsTemp(i) = (Arowptrs(i + 1) - Arowptrs(i)) + (Browptrs(i + 1) - Browptrs(i));
+  }
+  const Values& Avals;
```
See above note on Kokkos::View needing to be stored by value in functors, not by reference.

mhoemmen commented on packages/tpetra/core/ext/TpetraExt_MatrixMatrix_def.hpp:
```diff
@@ -895,6 +1080,687 @@ void Add(
+      inserted++;
+    }
+  }
+  const Values& Avals;
```
See above note on Kokkos::View needing to be stored by value in functors, not by reference.

mhoemmen commented on packages/tpetra/core/ext/TpetraExt_MatrixMatrix_def.hpp:
```diff
@@ -895,6 +1080,687 @@ void Add(
+      temp >>= 4;
+    }
+  }
+  //radixSort2: a CUDA-friendly equivalent for Tpetra::sort2
```
Tpetra already has a `KOKKOS_FUNCTION` version of sort; see Tpetra_Details_shortSort.hpp.  If we prefer radix sort, we should promote radix sort into its own header file and test it separately.

brian-kelley replied:
@mhoemmen I used the shellSort and it slowed down the sort+merge functor by about 65%, so I will move the radix sort into that file and document/test it.

mhoemmen commented on packages/tpetra/core/ext/TpetraExt_MatrixMatrix_def.hpp:
```diff
@@ -895,6 +1080,687 @@ void Add(
+    }
+    Crowcounts(i) = finalPos;
+  }
+  const RowPtrs& CrowptrTemp;
```
See above note on Kokkos::View needing to be stored by value in functors, not by reference.

mhoemmen commented on packages/tpetra/core/ext/TpetraExt_MatrixMatrix_def.hpp:
```diff
@@ -895,6 +1080,687 @@ void Add(
+      Cvals(Crowptr(i) + j) = CvalsOver(CrowptrTemp(i) + j);
+    }
+  }
+  const RowPtrs& Crowptr;
```
See above note on Kokkos::View needing to be stored by value in functors, not by reference.

mhoemmen commented on packages/tpetra/core/ext/TpetraExt_MatrixMatrix_def.hpp:
```diff
@@ -895,6 +1080,687 @@ void Add(
+  const typename AddKernels<SC, LO, GO, NO>::ColIndsArray& Bcolinds, 
+  const typename AddKernels<SC, LO, GO, NO>::ISC scalarB,
+  GO numGlobalCols,
+  Teuchos::RCP<typename AddKernels<SC, LO, GO, NO>::ValuesArray>& Cvals,
```
See above note on Kokkos::View needing to be stored by value in functors, not by reference.  Also, see above note on RCP vs. Kokkos::View (short answer: just store Kokkos::View; forget about RCP).

mhoemmen commented on packages/tpetra/core/ext/TpetraExt_MatrixMatrix_def.hpp:
```diff
@@ -895,6 +1080,687 @@ void Add(
+  const typename AddKernels<SC, LO, GO, NO>::GlobalColIndsArray& Bcolinds, 
+  const typename AddKernels<SC, LO, GO, NO>::ISC scalarB,
+  GO numGlobalCols,
+  Teuchos::RCP<typename AddKernels<SC, LO, GO, NO>::ValuesArray>& Cvals,
```
Don't use RCP; just pass ValuesArray, etc. by nonconst reference, since they are output arguments.

mhoemmen commented on packages/tpetra/core/test/MatrixMatrix/MatrixMatrix_UnitTests.cpp:
```diff
@@ -1140,6 +1142,238 @@ TEUCHOS_UNIT_TEST_TEMPLATE_4_DECL(Tpetra_MatMat, ATI_range_row_test, SC, LO, GO,
+      {
+        rowptrs[m][row] = row * nnzPerRow;
+      }
+      valsCRS[m] = rcp(new ValuesType("Values", nrows * nnzPerRow));
```
See above comments on RCP vs. View (short answer: no RCP).

brian-kelley commented:
@mhoemmen, I just pushed a change that addresses the easy things here (style, naming, locallySameAs, kernels in Details, remove RCP, remove references to View in functors, spelling). I still need to replace the radix sort and I haven't looked at the Tpetra Details prefix sum.

Could you clarify about the fused import/fillcomplete for CrsMatrix? I don't think any importing at all is necessary here since at that point Aprime, Bprime and C all have exactly the same rowmaps. I don't think any communication should be necessary at all here, except for computing global constants. All the sorting and merging has already been done in my kernels.

brian-kelley commented on packages/tpetra/core/ext/TpetraExt_MatrixMatrix_def.hpp:
```diff
@@ -479,6 +478,103 @@ void Add(
+  typedef typename AddKernels<Scalar, LocalOrdinal, GlobalOrdinal, Node>::ColIndsArray LView;
+  //Functors (explained in the procedural code below)
+  auto nentries = gids.dimension_0();
+  //each entry of entryUnion is 0 unless there is a local entry in that column (then it is 1)
```
@mhoemmen I'm thinking that the allocation time and extra memory usage here is acceptable, because rowptrs for each matrix is the same size (at least 3 of those exist during add()), and values/colinds are at least that big (even if only 1-2 entries per row).

brian-kelley commented on packages/tpetra/core/ext/TpetraExt_MatrixMatrix_def.hpp:
```diff
@@ -496,23 +592,27 @@ add (const Scalar& alpha,
+  typedef Import<LO,GO,NO>                  import_type;
+  typedef Export<LO,GO,NO>                  export_type;
+  typedef AddKernels<SC,LO,GO,NO>           AddKern;
+  std::string prefix_mmm = "TpetraExt::MatrixMatrix::add(): ";
```
@mhoemmen Should I replace all TEUCHOS_TEST_FOR_EXCEPTION in add() with this? I'm looking at MultiVector_def.hpp but it looks like TEUCHOS_TEST_FOR_EXCEPTION_CLASS_FUNC does the same thing.

brian-kelley replied:
@mhoemmen I did make this change here, but otherwise the std::string constructor time is negligible. On cee-compute (1 thread) 0.04 s was spent in std::string::string out of 1500 s total CPU time. This was with MMM_Timings off, but I assume users would have that off if they really care about performance.

mhoemmen commented:
@brian-kelley wrote:
> Could you clarify about the fused import/fillcomplete for CrsMatrix?

I'll take a closer look, but what I meant was that you're creating a Import from two Maps, rather than using the "union Import and column Maps" feature that would save some communication.

brian-kelley commented:
@mhoemmen I'm working on moving the sort over (I will squash that commit before this gets checked in)

mhoemmen commented:
@brian-kelley You might like to make a new header file, `Tpetra_Details_radixSort`.  I would prefer that if you don't mind; thanks!

brian-kelley commented:
@mhoemmen I was playing around with some different sorting algorithms and I have a bucket/insertion sort that beats the radix sort by about 10% on my machine, so I think that's the one I'll use on CPUs. I'll use the radix on GPUs though. The sorting is still by far the most expensive part of the unsorted kernel on low numbers of threads so it's the best thing to optimize now.

brian-kelley commented:
@mhoemmen I added the radix and bucket sort to new Details headers, and I am testing them in Sort/Sort.cpp in the same way as shellSort. I also added the execution_space::fence() calls where needed.

mhoemmen commented:
@brian-kelley Could you please send us an e-mail once you're done and ready for another review?  Sometimes GitHub notifications are weird.  Thanks!

mhoemmen commented:
Some weird thing happened this morning with the Kokkos CUDA build -- nothing to do with Tpetra -- I am trying to figure it out. Once I do that, I'll test this PR.

mhoemmen commented:
FYI @brian-kelley asked me to wait on the merge until he changes the code to use `Tpetra::Import::setUnion` (see my review above), and possibly also adds thread parallelization to the implementation of `setUnion`.

mhoemmen commented:
I'm doing CUDA + complex arithmetic testing now, just to make sure.

mhoemmen commented:
@brian-kelley There are some `Scalar` values going on (CUDA) device in functors.  If enabling complex arithmetic, those would be `std::complex`, which won't work on device.  Will you have a chance to fix those or shall I do it?

mhoemmen commented:
Here are the two warnings that told me about this.  It looks pretty easy to fix -- just replace `Scalar` with `impl_scalar_type` when instantiating the functors.
```
.../Trilinos/packages/tpetra/core/ext/TpetraExt_MatrixMatrix_def.hpp(1343): warning: calling a __host__ function("Kokkos::operator *<double> ") from a __host__ __device__ function("Tpetra::MatrixMatrix::AddUnsortedFunctors::UnmergedSumFunctor< ::std::complex<double> , int,  ::Kokkos::View< ::Kokkos::complex<double>  *,  ::Kokkos::LayoutRight,  ::Kokkos::Device< ::Kokkos::Cuda,  ::Kokkos::CudaUVMSpace> ,  ::Kokkos::MemoryTraits<(unsigned int)0u>  > ,  ::Kokkos::View<unsigned int *,  ::Kokkos::LayoutLeft,  ::Kokkos::Device< ::Kokkos::Cuda,  ::Kokkos::CudaUVMSpace> ,  ::Kokkos::MemoryTraits<(unsigned int)0u>  > ,  ::Kokkos::View<long long *,  ::Kokkos::Device< ::Kokkos::Cuda,  ::Kokkos::CudaUVMSpace>  > > ::operator ()") is not allowed

.../Trilinos/packages/tpetra/core/ext/TpetraExt_MatrixMatrix_def.hpp(1349): warning: calling a __host__ function("Kokkos::operator *<double> ") from a __host__ __device__ function("Tpetra::MatrixMatrix::AddUnsortedFunctors::UnmergedSumFunctor< ::std::complex<double> , int,  ::Kokkos::View< ::Kokkos::complex<double>  *,  ::Kokkos::LayoutRight,  ::Kokkos::Device< ::Kokkos::Cuda,  ::Kokkos::CudaUVMSpace> ,  ::Kokkos::MemoryTraits<(unsigned int)0u>  > ,  ::Kokkos::View<unsigned int *,  ::Kokkos::LayoutLeft,  ::Kokkos::Device< ::Kokkos::Cuda,  ::Kokkos::CudaUVMSpace> ,  ::Kokkos::MemoryTraits<(unsigned int)0u>  > ,  ::Kokkos::View<long long *,  ::Kokkos::Device< ::Kokkos::Cuda,  ::Kokkos::CudaUVMSpace>  > > ::operator ()") is not allowed
```

mhoemmen commented:
Hm, I'm getting some build errors too:
```
[ 94%] Building CXX object packages/tpetra/core/test/Sort/CMakeFiles/TpetraCore_Sort.dir/Sort.cpp.o
.../Trilinos/packages/tpetra/core/test/Sort/Sort.cpp(256): error: no instance of function template "Tpetra::Details::radixSortKeysAndValues" matches the argument list
            argument types are: (short [2], short [2], double [2], double [2], int)
          detected during:
            instantiation of "void <unnamed>::test_fixedTypes_fixedArrayLength<KeyType,ValueType,arrayLength>(__nv_bool &, Teuchos::FancyOStream &, const KeyType *, const ValueType *, __nv_bool, __nv_bool) [with KeyType=short, ValueType=double, arrayLength=2]"
(315): here
            instantiation of "void <unnamed>::test_fixedTypes<KeyType,ValueType>(__nv_bool &, Teuchos::FancyOStream &, const char *, const char *) [with KeyType=short, ValueType=double]"
```
and many more like this, for other types.

brian-kelley commented:
@mhoemmen I just pushed a fix to these issues.

mhoemmen commented:
Thanks @brian-kelley ! :-D  I'll try again tomorrow morning (I'm finishing up some left-over CrsMatrix (un)pack stuff tonight -- I want CrsMatrix to implement the "new" DistObject interface).

mhoemmen commented:
@brian-kelley May I squash the first two commits into an atomic commit, and leave the TpetraExt add() test commit separate?

brian-kelley commented:
@mhoemmen Yes.

mhoemmen commented:
How am I supposed to test CUDA builds with #1726?  Argh....

mhoemmen commented:
@brian-kelley I squashed all three commits together into one, for better atomicity.  I'm running the remote check-in test script now.

mhoemmen commented:
@brian-kelley It looks like these changes make an Ifpack2 test fail:
```
$ ctest -V -R Ifpack2_RILUKSing
...
26: 0. Ifpack2RILUKSingleProcess_double_int_int_Test0_UnitTest ... [Passed] (0.00768 sec)
26: 1. Ifpack2RILUKSingleProcess_double_int_int_Test1_UnitTest ... [Passed] (0.0065 sec)
26: 2. Ifpack2RILUKSingleProcess_double_int_int_FillLevel_UnitTest ...
26:  Ifpack2::RILUK: FillLevel
26:  reading Lfactor_bw2.mm, Ufactor_bw2.mm, Dfactor_bw2.mm
26:  bandwidth = 2, lof = 0
26:
26:  p=0: *** Caught standard std::exception of type 'Teuchos::NullReferenceError' :
26:
26:   .../Trilinos/packages/teuchos/core/src/Teuchos_RCPNode.cpp:758:
26:
26:   Throw number = 10
26:
26:   Throw test that evaluated to true: true
26:
26:   Teuchos::RCP<Tpetra::Import<int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> > const> : You can not call operator->() or operator*() if getRawPtr()==0!
26:  [FAILED]  (0.0089 sec) Ifpack2RILUKSingleProcess_double_int_int_FillLevel_UnitTest
26:  Location: .../Trilinos/packages/ifpack2/test/unit_tests/Ifpack2_UnitTestSingleProcessRILUK.cpp:277
26:
26: 3. Ifpack2RILUKSingleProcess_double_int_int_IgnoreRowMapGIDs_UnitTest ... [Passed] (0.0572 sec)
26: 4. Ifpack2RILUKSingleProcess_double_int_int_TestGIDConsistency_UnitTest ... [Passed] (0.00241 sec)
26: 5. Ifpack2RILUKSingleProcess_double_int_longlong_Test0_UnitTest ... [Passed] (0.00392 sec)
26: 6. Ifpack2RILUKSingleProcess_double_int_longlong_Test1_UnitTest ... [Passed] (0.00567 sec)
26: 7. Ifpack2RILUKSingleProcess_double_int_longlong_FillLevel_UnitTest ...
26:  Ifpack2::RILUK: FillLevel
26:  reading Lfactor_bw2.mm, Ufactor_bw2.mm, Dfactor_bw2.mm
26:  bandwidth = 2, lof = 0
26:
26:  p=0: *** Caught standard std::exception of type 'Teuchos::NullReferenceError' :
26:
26:   .../Trilinos/packages/teuchos/core/src/Teuchos_RCPNode.cpp:758:
26:
26:   Throw number = 43
26:
26:   Throw test that evaluated to true: true
26:
26:   Teuchos::RCP<Tpetra::Import<int, long long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> > const> : You can not call operator->() or operator*() if getRawPtr()==0!
26:  [FAILED]  (0.00759 sec) Ifpack2RILUKSingleProcess_double_int_longlong_FillLevel_UnitTest
26:  Location: .../Trilinos/packages/ifpack2/test/unit_tests/Ifpack2_UnitTestSingleProcessRILUK.cpp:277
26:
26: 8. Ifpack2RILUKSingleProcess_double_int_longlong_IgnoreRowMapGIDs_UnitTest ... [Passed] (0.0568 sec)
26: 9. Ifpack2RILUKSingleProcess_double_int_longlong_TestGIDConsistency_UnitTest ... [Passed] (0.00241 sec)
26:
26: The following tests FAILED:
26:     2. Ifpack2RILUKSingleProcess_double_int_int_FillLevel_UnitTest ...
26:     7. Ifpack2RILUKSingleProcess_double_int_longlong_FillLevel_UnitTest ...
```

mhoemmen commented:
@brian-kelley btw I pushed the squashed version of this PR to develop.  Thanks!

Here is the squashed commit:

https://github.com/trilinos/Trilinos/commit/2e7cba859bafd69602ad31c1e8f0f5c9b7f3b443
